jisantuc/panrec
# Write records

Overview
-----

This PR figures out how to write collections of records to disk. It does not solve the problem of parsing several records of a type from a file. If I solved that problem the CLI would work.

It makes it so that the following all basically work, though there's no translation from `Primitive`s back to types that the source language understands (that's another function that needs to exist on `Record` that doesn't exist yet). It also makes some pretty strong assumptions about how higher-kinded types are supposed to look when written -- i.e., an optional type is written as `Option[Foo]` instead of typescript's style `Option<Foo>` or haskell's `Option Foo`. Figuring out a way to specify that style is also a problem for future me. That dude has all kinds of time.

What it makes possible is:

```haskell
> let foo = Class [("foo", Double')] "Foo"
> let bar = Class [("bar", String')] "Bar"
> writeRecords "example.ts" [foo, bar]
```
--> 

```typescript
class Foo{
    constructor(foo: Double) {
        this.foo = foo;
    }
}

class Bar{
    constructor(bar: String) {
        this.bar = bar;
    }
}
```

Conversion:

```haskell
> let scFoo = fromRecord foo :: CaseClass
> let scBar = fromRecord bar :: CaseClass
> writeRecords "example.sc" [scFoo, scBar]
```

-->

```scala
case class Foo(foo: Double)

case class Bar(bar: String)
```

Conversion back:

```
> let tsFoo = fromRecord scFoo :: Class
> let tsBar = fromRecord scBar :: Class
> writeRecords "example2.ts" [tsFoo, tsBar]
```

-->

```typescript
class Foo{
    constructor(foo: Double) {
        this.foo = foo;
    }
}

class Bar{
    constructor(bar: String) {
        this.bar = bar;
    }
}
```

Notes
-----

Handling of higher-kinded types in typescript was also left out for now -- handling options as `primParser <* string "| null"` was an infinite recursion, but I can't have another prim parser succeed and _then_ discover that I actually wanted an `Option`, so I think the solution here is to demand that typescript input files have type aliases defined e.g. for `Option<Foo>` or `Maybe<Foo>`. It's not ideal and puts a lot on the consumer but I'm not sure how to break the infinite recursion otherwise. Higher-kinded types from Scala work great though:

```haskell
> SC.parseField "bar: Option[Option[Either[IO[Int], Either[Int, String]]]]\n" 
Right ("bar",Option[Option[Either[IO[Int], Either[Int, String]]]])
```

Recursing with monads is a blast :sunglasses: 

Testing
-----

- run tests
- confirm file output works from ghci
